routing_algorithms/shortest_path.hpp

```diff
@@ -51,294 +53,482 @@ class ShortestPathRouting final
+                                    source_phantom.GetForwardWeightPlusOffset(),
+                                source_phantom.forward_node_id);
+        }
+        if (search_from_reverse_node)
+        {
+            forward_heap.Insert(source_phantom.reverse_node_id,
+                                total_distance_to_reverse -
+                                    source_phantom.GetReverseWeightPlusOffset(),
+                                source_phantom.reverse_node_id);
+        }
+        if (search_to_forward_node)
+        {
+            reverse_heap.Insert(target_phantom.forward_node_id,
+                                target_phantom.GetForwardWeightPlusOffset(),
+                                target_phantom.forward_node_id);
+        }
+        if (search_to_reverse_node)
+        {
+            reverse_heap.Insert(target_phantom.reverse_node_id,
+                                target_phantom.GetReverseWeightPlusOffset(),
+                                target_phantom.reverse_node_id);
+        }
+        BOOST_ASSERT(forward_heap.Size() > 0);
+        BOOST_ASSERT(reverse_heap.Size() > 0);
+        super::Search(forward_heap, reverse_heap, new_total_distance, leg_packed_path);
+    }
 
-        engine_working_data.InitializeOrClearFirstThreadLocalStorage(
-            super::facade->GetNumberOfNodes());
-        engine_working_data.InitializeOrClearSecondThreadLocalStorage(
-            super::facade->GetNumberOfNodes());
-        engine_working_data.InitializeOrClearThirdThreadLocalStorage(
-            super::facade->GetNumberOfNodes());
+    // If source and target are reverse on a oneway we need to find a path
+    // that connects the two. This is _not_ the shortest path in our model,
+    // as source and target are on the same edge based node.
+    // We force a detour by inserting "virtaul vias", which means we search a path
+    // from all nodes that are connected by outgoing edges to all nodes that are connected by
+    // incoming edges.
+    // ------^
+    // |     ^source
+    // |     ^
+    // |     ^target
+    // ------^
+    void SearchLoop(QueryHeap &forward_heap,
+                QueryHeap &reverse_heap,
+                const bool search_from_forward_node,
+                const bool search_from_reverse_node,
+                const bool search_to_forward_node,
+                const bool search_to_reverse_node,
+                const PhantomNode &source_phantom,
+                const PhantomNode &target_phantom,
+                const int total_distance_to_forward,
+                const int total_distance_to_reverse,
+                int &new_total_distance_to_forward,
+                int &new_total_distance_to_reverse,
+                std::vector<NodeID> &leg_packed_path_forward,
+                std::vector<NodeID> &leg_packed_path_reverse) const
+    {
+        // silence unused warning
+        (void) search_from_forward_node;
+        (void) search_from_reverse_node;
```

Comment:
Might be neater to do this in the signature instead:

    const bool /* search_from_forward_node */,
    const bool /* search_from_reverse_node */,

Why are these unused?